### PR TITLE
Fix potential connection leak in Raptor ShardIterator

### LINKED ISSUE #5762: Raptor - connection leak
When using MySQL as backend there is a connection leak:

```
2016-07-28T14:32:56.300Z        WARN    HikariPool-1 housekeeper        com.zaxxer.hikari.pool.ProxyLeakTask    Connection leak detection triggered for com.mysql.jdbc.JDBC4Connection@7ef9c8a5, stack trace follows
java.lang.Exception: Apparent connection leak detected
        at org.skife.jdbi.v2.DBI.open(DBI.java:212)
        at com.facebook.presto.raptor.metadata.ShardIterator.<init>(ShardIterator.java:87)
        at com.facebook.presto.raptor.metadata.DatabaseShardManager.getShardNodes(DatabaseShardManager.java:460)
        at com.facebook.presto.raptor.RaptorSplitManager$RaptorSplitSource.<init>(RaptorSplitManager.java:164)
        at com.facebook.presto.raptor.RaptorSplitManager.getSplits(RaptorSplitManager.java:107)
        at com.facebook.presto.split.SplitManager.getSplits(SplitManager.java:45)
        at com.facebook.presto.sql.planner.DistributedExecutionPlanner$Visitor.visitTableScan(DistributedExecutionPlanner.java:112)
        at com.facebook.presto.sql.planner.DistributedExecutionPlanner$Visitor.visitTableScan(DistributedExecutionPlanner.java:92)
        at com.facebook.presto.sql.planner.plan.TableScanNode.accept(TableScanNode.java:135)
        at com.facebook.presto.sql.planner.DistributedExecutionPlanner$Visitor.visitProject(DistributedExecutionPlanner.java:228)
        at com.facebook.presto.sql.planner.DistributedExecutionPlanner$Visitor.visitProject(DistributedExecutionPlanner.java:92)
        at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:81)
        at com.facebook.presto.sql.planner.DistributedExecutionPlanner.plan(DistributedExecutionPlanner.java:78)
        at com.facebook.presto.sql.planner.DistributedExecutionPlanner.plan(DistributedExecutionPlanner.java:83)
        at com.facebook.presto.sql.planner.DistributedExecutionPlanner.plan(DistributedExecutionPlanner.java:83)
        at com.facebook.presto.sql.planner.DistributedExecutionPlanner.plan(DistributedExecutionPlanner.java:83)
        at com.facebook.presto.sql.planner.DistributedExecutionPlanner.plan(DistributedExecutionPlanner.java:83)
        at com.facebook.presto.sql.planner.DistributedExecutionPlanner.plan(DistributedExecutionPlanner.java:83)
        at com.facebook.presto.sql.planner.DistributedExecutionPlanner.plan(DistributedExecutionPlanner.java:83)
        at com.facebook.presto.sql.planner.DistributedExecutionPlanner.plan(DistributedExecutionPlanner.java:83)
        at com.facebook.presto.execution.SqlQueryExecution.planDistribution(SqlQueryExecution.java:303)
        at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:226)
        at com.facebook.presto.execution.QueuedExecution.lambda$start$1(QueuedExecution.java:62)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)

```
I guess that the following code from ShardIterator.java should be catching Exception not SqlException
```
        try {
            connection = dbi.open().getConnection();
            statement = connection.prepareStatement(sql);
            enableStreamingResults(statement);
            predicate.bind(statement);
            log.debug("Running query: %s", statement);
            resultSet = statement.executeQuery();
        }
        catch (SQLException e) {
            close();
            throw metadataError(e);
        }

### PR DESCRIPTION
**Pull request by electrum**
Fixes #5762

### CODE REVIEW

File: presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardIterator.java [hunk 1]
```diff
@@ -95,6 +95,10 @@ public ShardIterator(
             close();
             throw metadataError(e);
         }
+        catch (Throwable t) {
+            close();
```
Comment (erichwang) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardIterator.java]: why don't you just put the close() in a finally block?

Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardIterator.java]: It should only close on failure.

**Comment by electrum** [comment]
@erichwang suggested the following structure
```java
boolean success = false;
try {
    ...
    success = true;
}
catch (SQLException e) {
    throw metadataError(e);
}
finally {
    if (!success) {
        close();
    }
}

**Comment by martint** [comment]
I think the second form is more explicit about what's going on. The first one has the disadvantage that you have to do control flow calculations in your head to realize that the finally block only really executes in case an exception is thrown.

**Comment by martint** [comment]
looks good

**Pull request closed without merging**